onnxruntime/core/graph/graph.cc

```diff
@@ -2445,6 +2445,50 @@ void Graph::CleanUnusedInitializers() {
                 });
 }
 
+static bool ModelUseFP16Helper(const onnx::TypeProto& type_proto) {
+  switch (type_proto.value_case()) {
+    case ::onnx::TypeProto::ValueCase::kTensorType: {
+      if (type_proto.has_tensor_type()) {
+        auto& tensor_type = type_proto.tensor_type();
+        if (tensor_type.elem_type() == ONNX_NAMESPACE::TensorProto_DataType::TensorProto_DataType_FLOAT16) {
+          return true;
+        }
+      }
+      break;
+    }
+    case ::onnx::TypeProto::ValueCase::kSequenceType: {
+      if (type_proto.has_sequence_type()) {
+        auto& sequence_type = type_proto.sequence_type();
+        return ModelUseFP16Helper(sequence_type.elem_type());
+      }
+      break;
+    }
+    case ::onnx::TypeProto::ValueCase::kMapType: {
+      if (type_proto.has_map_type()) {
+        auto& map_type = type_proto.map_type();
+        return ModelUseFP16Helper(map_type.value_type());
+      }
+      break;
+    }
+    default:
+      break;
+  }
+  return false;
+}
+
+Status Graph::DoInputsHaveFP16()
+{
+  auto& inputs = graph_proto_->input();
+  for (auto& input : inputs) {
+    if (input.has_name() && input.has_type() && ModelUseFP16Helper(input.type())) {
+      inputsHaveFp16_ = true;
+      return Status::OK();
+    }
+  }
+  inputsHaveFp16_ = false;
```

Comment:
>inputsHaveFp16_ = false; [](start = 2, length = 24)

This should be done at GraphIntialization time so the method is const

Reply:
I put it in the constructor. Does that look good?

---
In reply to: [382318426](https://github.com/microsoft/onnxruntime/pull/3061#discussion_r382318426) [](ancestors = 382318426)